infor-design/enterprise
# ng#467 - Add API to control Buttons inside Modals

**Explain the _details_ for making this change. What existing problem does the pull request solve?**
This PR creates a "ButtonSet" API that can be used within Modals to control an entire row of Modal Buttons at once. To add this functionality, this PR:
- Fleshes out many aspects of the existing Button API to be more programmatic and controlled with Javascript API.
- Takes the previous syntax for defining a Button from the Modal component's `buttons` setting and translates it to a new component, `ButtonSet`, for adding/removing buttons by using the element, API reference, or an ID.
- Replaces the existing method of dealing with Modal Buttons with the new ButtonSet component, allowing it to be controlled from the Modal API.

**Related github/jira issue (required)**:
Related to infor-design/enterprise-ng#467

**Steps necessary to review your pull request (required)**:
Pull this branch, build, and run the demoapp.  Then:

_To test the ButtonSet and new Button settings:_
- Open http://localhost:4000/components/button/test-buttonset-gauntlet.html
- Use all the various settings to test adding buttons.
- Use `reset` to clear the ButtonSet.
- Use the "Default/Modal" switch to toggle styles.  Make sure the background color disappears on Primary buttons when using 'modal` mode.

_To test Modal integration:_
- Basically [smoke test all Modal samples](http://localhost:4000/components/modal/list) and make sure they match what's currently on `master`.  These are all using the new component, but their general look/feel/actions should be the same.
- Also check that all [Contextual Action Panel samples](http://localhost:4000/components/contextualactionpanel/list) work as expected.  These are not using the new component and continue to use either standard Toolbar/Flex Toolbar for now.  Nothing should have changed here.
- There are also many new functional tests for Button, ButtonSet, and Modal that cover the button definition cases described in the ticket.

**Included in this Pull Request**:
